src/Data/Parameterized/TH/GADT.hs

```diff
@@ -493,6 +497,151 @@ showCon p nm n = do
 matchShowCtor :: ExpQ -> ConstructorInfo -> MatchQ
 matchShowCtor p con = showCon p (constructorName con) (length (constructorFields con))
 
+-- | Generate a "repr" or singleton type from a data kind. For nullary
+-- constructors, this works as follows:
+--
+-- @
+-- data T1 = A | B | C
+-- \$(mkRepr ''T1)
+-- ======>
+-- data T1Repr (tp :: T1)
+--   where
+--     ARepr :: T1Repr 'A
+--     BRepr :: T1Repr 'B
+--     CRepr :: T1Repr 'C
+-- @
+--
+-- For constructors with fields, we assume each field type @T@ already has a
+-- corresponding repr type @TRepr :: T -> *@.
+--
+-- @
+-- data T2 = T2_1 T1 | T2_2 T1
+-- \$(mkRepr ''T2)
+-- ======>
+-- data T2Repr (tp :: T2)
+--   where
+--     T2_1Repr :: T1Repr tp -> T2Repr ('T2_1 tp)
+--     T2_2Repr :: T1Repr tp -> T2Repr ('T2_2 tp)
+-- @
+--
+-- Constructors with multiple fields work fine as well:
+--
+-- @
+-- data T3 = T3 T1 T2
+-- \$(mkRepr ''T3)
+-- ======>
+-- data T3Repr (tp :: T3)
+--   where
+--     T3Repr :: T1Repr tp1 -> T2Repr tp2 -> T3Repr ('T3 tp1 tp2)
+-- @
+--
+-- This is generally compatible with other "repr" types provided by
+-- @parameterized-utils@, such as @NatRepr@ and @PeanoRepr@:
+--
+-- @
+-- data T4 = T4_1 Nat | T4_2 Peano
+-- \$(mkRepr ''T4)
+-- ======>
+-- data T4Repr (tp :: T4)
+--   where
+--     T4Repr :: NatRepr tp1 -> PeanoRepr tp2 -> T4Repr ('T4 tp1 tp2)
+-- @
+mkRepr :: Name -> DecsQ
```

Comment:
I'm not sure how robust this function is with respect to polymorphism. What will it do on a data type like this, for example?

```hs
data Foo a b = MkFoo a b
```

Reply:
Oh, I see now. Thanks! `parameterized-utils` does not have anything analogous to `Sing`, and I don't think we have a need for it necessarily; most of the use cases I've seen involve plain, basically monomorphic data kinds (the types you see in `crucible` and `what4` are typically as complex as it gets).

I did not think very hard about generating good error messages, so I need to put a bit more effort into that. I think we basically want a check for data kinds that makes sure they are totally monomorphic, and it might also be a good idea to make sure that all the types they are built from also have corresponding `repr` types.

Reply:
I suppose an alternative approach would be to do something like what the [`List` repr does](http://hackage.haskell.org/package/parameterized-utils-2.1.3.0/docs/Data-Parameterized-List.html#t:List) and parameterize it by another type variable `f`:

```hs
data List :: (k -> Type) -> [k] -> Type where
  Nil  :: List f '[]
  (:<) :: f tp -> List f tps -> List f (tp : tps)
```

You'd need a type variable like `f` for each type parameter your original data type has.